elks/arch/i86/kernel/irq.c

```diff
@@ -21,75 +21,85 @@
 #include <linuxmt/sched.h>
 #include <linuxmt/timer.h>
 #include <linuxmt/types.h>
+#include <linuxmt/heap.h>
 
 #include <arch/ports.h>
+#include <arch/segment.h>
+#include <arch/irq.h>
 
-/* Enable 80386 Traps */
-/*#define ENABLE_TRAPS */
-
-struct irqaction {
-	void (*handler)();
-	void *dev_id;
-};
-
-static void default_handler(int i, void *regs, void *dev);
-
-static struct irqaction irq_action[] = {
-    {default_handler, NULL}, {default_handler, NULL}, {default_handler, NULL},
-    {default_handler, NULL}, {default_handler, NULL}, {default_handler, NULL},
-    {default_handler, NULL}, {default_handler, NULL}, {default_handler, NULL},
-    {default_handler, NULL}, {default_handler, NULL}, {default_handler, NULL},
-    {default_handler, NULL}, {default_handler, NULL}, {default_handler, NULL},
-    {default_handler, NULL},
-#ifdef ENABLE_TRAPS
-    {default_handler, NULL}, {default_handler, NULL}, {default_handler, NULL},
-    {default_handler, NULL}, {default_handler, NULL}, {default_handler, NULL},
-    {default_handler, NULL}, {default_handler, NULL}, {default_handler, NULL},
-    {default_handler, NULL}, {default_handler, NULL}, {default_handler, NULL},
-    {default_handler, NULL}, {default_handler, NULL}, {default_handler, NULL},
-    {default_handler, NULL},
-#endif
-};
+
+static irq_handler irq_action [16];
 
 /*
  *	Called by the assembler hooks
  */
 
 void do_IRQ(int i,void *regs)
 {
-    register struct irqaction *irq = irq_action + i;
-
-    irq->handler(i, regs, irq->dev_id);
+    register irq_handler ih = irq_action [i];
+    if (!ih) {
+        printk("Unexpected interrupt: %u\n", i);
+        return;
+    }
+	(*ih)(i, regs);
 }
 
-static void default_handler(int i, void *regs, void *dev)
-{
-#ifdef ENABLE_TRAPS
-    if (i > 15)
-	printk("Unexpected trap: %u\n", i-16);
-    else
-#endif
-	printk("Unexpected interrupt: %u\n", i);
-}
 
-int request_irq(int irq, void (*handler)(int,struct pt_regs *,void *), void *dev_id)
+// TODO: simplify the whole by replacing IRQ by INT number
+// Would also allow to handle any of the 0..255 interrupts
+// including the 0..7 processor exceptions & traps
+
+struct int_handler {
+	byte_t call;  // CALLF opcode (9Ah)
+	int_proc proc;
+	word_t seg;
+	byte_t irq;
+} __attribute__ ((packed));
+
+typedef struct int_handler int_handler_s;
+
+
+// Add a dynamically allocated handler
+// that redirects to the generic handler
+
+static int int_handler_add (int irq, int vect, int_proc proc)
+	{
+	int_handler_s * h = (int_handler_s *) heap_alloc (sizeof (int_handler_s), HEAP_TAG_INTHAND);
+	if (!h) return -ENOMEM;
+
+	h->call = 0x9A;  // CALLF opcode
+	h->proc = proc;
+	h->seg  = int_vector_set (vect, (int_proc) h, kernel_ds);
```

Comment:
Now that I finally fully understand the new _irqit code, I think a change should be made here to clean up the hack in int_vector_set of returning CS. This hard-to-find result actually delayed my understanding of your new approach.

But the real reason is that now that you understand how kernel medium model works, returning CS has no business in int_vector_set. Change that function to be a void proc and set h->seg = kernel_cs here. In this way, system won't be dependent on ASM routine in the case we want to move _irqit to different segment.

Reply:
Agree, returning CS is useless. Will remove that.

Reply:
Removed in 5adc931